File: src/two_crystal_balls.rs
use std::usize;

pub fn two_crystal_balls(breaks: &[bool]) -> Option<usize> {
    let n = breaks.len();
    let step = (n as f64).sqrt().floor() as usize;

    let i = (step..n).step_by(step).find(|&i| breaks[i]).unwrap_or(n);

    for j in (i - step)..i {
        if breaks[j] {
            return Some(j as usize)
        }
    }

    None
}

File: two-crystal-balls/two_crystal_balls.go
package two_crystal_balls

import (
	"errors"
	"fmt"
	"math"
)

func two_crystal_balls(breaks []bool) (int, error) {
	n := len(breaks)
	step := int(math.Floor(math.Sqrt(float64(n))))

	j := step
	for ; j < n; j += step {
		if breaks[j] {
			break
		}
	}

	for i := j - step; i < j; i++ {
		if breaks[i] {
			return i, nil
		}
	}

	fmt.Printf("%d", j)
	return 0, errors.New("no breaking point in the list")
}
